near-plugins-derive/src/access_control_role.rs

```diff
@@ -176,7 +176,7 @@ pub fn derive_access_control_role(input: TokenStream) -> TokenStream {
             }
         }
 
-        ::#cratename::bitflags::bitflags! {
```

Comment:
Out of curiosity, what is the error message?

Reply:
With leading `::` there are a bunch of errors and I think the crucial one is:

```
could not find `near_plugins` in the list of imported crates
```

My gut feeling is that's due to `mod tests` in `near-plugins/src/pausable.rs` having

```rust
use crate as near-plugins;
```

instead of importing `near-plugins` in `Cargo.toml` (which would mean importing itself).

Reply:
**TL;DR**: I think it’s better to remove the leading `::`, it shouldn’t have been there.

**Background**:

During plugin macro expansion `#cratename` is resolved like this:

```rust
// Assume this is inside `quoute!`
#cratename::events::AsEvent::emit(&emit);

// It will be resolved to
near_plugins::events::AsEvent::emit(&emit);
```

So far only Acl plugins were referencing `#cratename` with leading `::`. It didn’t lead to any compilation errors as Acl wasn’t used in unit tests in `near-plugins`. This is changed in the current PR ([here](https://github.com/mooori/near-plugins/blob/56e595ea0e3d3d80008e21a651098de07040e2aa/near-plugins/src/pausable.rs#L94)).

Other plugins were using `#cratename` without leading `::`, for example:

https://github.com/aurora-is-near/near-plugins/blob/b2aa6a0ace5855688cad9b8824c37c6bf2d9e66e/near-plugins-derive/src/ownable.rs#L60

So removing the leading `::` makes usage of `#cratename` consistent - maybe I shouldn’t have sneaked it in here like this. Also near-sdk does _not_ use leading `::` when referring to itself inside `quote!` ([example](https://github.com/near/near-sdk-rs/blob/cb91d2513666bf41e0cbf8255b3645730d755dfe/near-sdk-macros/src/lib.rs#L602-L604)).

Reply:
> creating a alias for crate seems like something of an anti-pattern

Agreed - when following through with #48 those aliases will be removed since unit tests for plugin expansion in near-plugins will be replaced by integration tests. As long as these unit tests are still there, I'm afraid the alias is required to make macro expansion work.